### PR TITLE
Add support for SG, Volume Types, AZ and fix for disk indexing

### PR DESCRIPTION
**Pull request by dswebbthg**
4 Optional Flags added
- Added security group support during port creation
- Volume type for block devices (if not supplied will default to OS default volume type)
- Availibility zones for the virtual machine and block device (single mapping for both).
- The ability to disable running virt-v2v-inplace (defaults to running it)
- The ability to define virtio-scsi and virtio for disk device types

1 Change:
- Moved volume metadata updates to map prior to volume creation.

1 Bug fix:

- Added disk index for multi disk migrations as without it the server creation would fail.
- Changed WriteChangeID to only update the changeID.Value as otherwise it would fully replace all volume metadata with only the fields in itself.

### CODE REVIEW

**Review by mnaser — CHANGES_REQUESTED** [review]
Awesome.  There are a few things that would make things neater.  If you want me to pick this up, I'll be glad to.

**Comment by cohl790** [comment]
> 4 Optional Flags added
> 
> * Added security group support during port creation
> * Volume type for block devices (if not supplied will default to OS default volume type)
> * Availibility zones for the virtual machine and block device (single mapping for both).
> * The ability to disable running virt-v2v-inplace (defaults to running it)
> * The ability to define virtio-scsi and virtio for disk device types
> 
> 1 Change:
> 
> * Moved volume metadata updates to map prior to volume creation.
> 
> 1 Bug fix:
> 
> * Added disk index for multi disk migrations as without it the server creation would fail.

Having the option for volume type would be a great enhancement.

File: main.go [hunk 1]
```diff
@@ -230,12 +258,22 @@ func init() {
...
+
+	rootCmd.PersistentFlags().StringVar(&busType, "disk-bus-type", "virtio", "Specifies the type of disk controller to attach disk devices to.")
+
 	cutoverCmd.Flags().StringVar(&flavorId, "flavor", "", "OpenStack Flavor ID")
 	cutoverCmd.MarkFlagRequired("flavor")
 
 	cutoverCmd.Flags().Var(&networkMapping, "network-mapping", "Network mapping (e.g. 'mac=00:11:22:33:44:55,network-id=6bafb3d3-9d4d-4df1-86bb-bb7403403d24,subnet-id=47ed1da7-82d4-4e67-9bdd-5cb4993e06ff[,ip=1.2.3.4]')")
 	cutoverCmd.MarkFlagRequired("network-mapping")
 
+	cutoverCmd.Flags().StringVar(&securityGroups, "security-groups", "", "Openstack security groups, comma separated (e.g. '42c5a89e-4034-4f2a-adea-b33adc9614f4,6647122c-2d46-42f1-bb26-f38007730fdc')")
```
[the remark this answers is not in the document]
Comment (dswebbthg) [reply main.go]: Nice, I didn't see that was an option.

File: internal/target/openstack.go [hunk 2]
```diff
@@ -63,16 +69,32 @@ func findDevice(volumeID string) (string, error) {
...
+		"vm":          t.VirtualMachine.Reference().Value,
+		"disk":        t.Disk.DiskObjectId,
+	}
+
+	opts := ctx.Value("volumeCreateOpts").(*VolumeCreateOpts)
+
+	if opts.BusType == "scsi" {
+		volumeMetadata["hw_disk_bus"] = "scsi"
+		volumeMetadata["hw_scsi_model"] = "virtio-scsi"
+	}
```
[the remark this answers is not in the document]
Comment (dswebbthg) [reply internal/target/openstack.go:83]: I'm not sure tbh.  I'm assuming you could query VMWare to look at it's disk controllers but I'm not sure how that would be done.

File: main.go [hunk 3]
```diff
@@ -230,12 +258,22 @@ func init() {
 	rootCmd.PersistentFlags().StringVar(&path, "vmware-path", "", "VMware VM path (e.g. '/Datacenter/vm/VM')")
 	rootCmd.MarkPersistentFlagRequired("vmware-path")
 
+	rootCmd.PersistentFlags().StringVar(&availabilityZone, "availability-zone", "", "Openstack availability zone for blockdevice & server")
+
+	rootCmd.PersistentFlags().StringVar(&volumeType, "volume-type", "", "Openstack volume type")
+
+	rootCmd.PersistentFlags().StringVar(&busType, "disk-bus-type", "virtio", "Specifies the type of disk controller to attach disk devices to.")
```
[the remark this answers is not in the document]
Comment (dswebbthg) [reply main.go]: done

File: internal/target/openstack.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (michael-salo) [reply internal/target/openstack.go:83]: yeah, it's not something that can be automatically detect from the source VM.

[the remark this answers is not in the document]
Comment (mnaser) [reply internal/target/openstack.go:83]: Ok, got it, will mark as resolved then.

**Review by mnaser — APPROVED** [review]

**Comment by mnaser** [comment]
@dswebbthg This looks good to me at this points, am I good to proceed with merging or you have a few more things to add?

**Review by michael-salo — APPROVED** [review]
LGTM

**Pull request merged**